### PR TITLE
Add Unmatchable token, fixes #23

### LINKED ISSUE #23: Occasional failures of prop_commonDirectory
See the failing CI in #22.

The issue seems to be with cases where `[.]` appears somewhere other than at the beginning or directly after a path separator; the property is testing that the two patterns are equal, and they ought to behave the same for matching, but one has the `.` inside a `LongLiteral` constructor, and the other has it as an `ExtSeparator` constructor.

I think there are two possible ways to fix this. We could either modify `compile` so that any pair of patterns such as `"foo."` and `"foo[.]"` do come out the same, presumably by changing "implicit" dots, i.e. dots inside `LongLiteral` constructors, into "explicit" ones, i.e. ones inside `ExtSeparator` constructors, in all cases (unless this would cause `.` or `..` to match the pattern where they otherwise would not). Alternatively, we could modify the tests to account for this possible discrepancy while testing. I think the first option sounds simpler, and also would make the behaviour of `==` on Patterns a little more intuitive.

### PR DESCRIPTION
**Pull request by hdgarrood**
Optimize unmatchable patterns to just [Unmatchable], i.e.
`compile "[.]"`.

Also strengthen tests for commonDirectory edge cases to ensure that
tests fail a little more reliably if this issue is present.

### CODE REVIEW

File: System/FilePath/Glob/Base.hs [hunk 1]
```diff
@@ -517,7 +519,7 @@ charRange opts zs =
 
 
 optimize :: Pattern -> Pattern
-optimize = liftP (fin . go)
+optimize = liftP (fin . checkUnmatchable . convertExtSeparators . go)
```
Comment (Deewiant) [inline System/FilePath/Glob/Base.hs]: Can't `convertExtSeparators` be done inside `go`? The way it searches for context that would be directly available in `go` seems somewhat wasteful, on the eyes if not also the CPU. I.e. something like:

```haskell
go (PathSeparator : Literal '.' : _) = [Unmatchable]
go (PathSeparator : ExtSeparator : xs) = PathSeparator : ExtSeparator : go xs
go (ExtSeparator : xs) = Literal '.' : go xs

-- handle the cases where the whole pattern starts with a now-literalized [.]
checkUnmatchable (Literal '.' : _) = [Unmatchable]
checkUnmatchable (LongLiteral _ ('.' : _) : _) = [Unmatchable]
checkUnmatchable ts = if Unmatchable `elem` ts then [Unmatchable] else ts
```

File: tests/Tests/Instances.hs [hunk 2]
```diff
@@ -42,7 +42,8 @@ prop_monoidLaw3 opt x =
     in isRight e && mappend a mempty == a
 
 -- mappending two Patterns should be equivalent to appending the original
--- strings they came from and compiling that
+-- strings they came from and compiling that, as long as the second doesn't
+-- start with [.]
```
Comment (Deewiant) [inline tests/Tests/Instances.hs]: ...and those four other excluded cases that this comment doesn't mention. Maybe it's more misleading to mention one case out of five than to not mention any? Just musing, no big deal.

File: System/FilePath/Glob/Base.hs (hunk 1 above)
Comment (hdgarrood) [reply System/FilePath/Glob/Base.hs]: Oh of course, I'm not sure why I didn't think of that.

File: tests/Tests/Instances.hs (hunk 2 above)
Comment (hdgarrood) [reply tests/Tests/Instances.hs]: Good point, I'll remove it.

File: System/FilePath/Glob/Base.hs (hunk 1 above)
Comment (hdgarrood) [reply System/FilePath/Glob/Base.hs]: With this code, `go` will convert `ExtSeparator` into `Literal '.'` at the very start of a pattern, and then `checkUnmatchable` will convert that into just `[Unmatchable]`, so any pattern starting with a dot, whether explicit or not, becomes `[Unmatchable]`. I think we do need a separate function to deal with `ExtSeparator` at the very start, because the way `go` is used recursively means that we are not able to tell when we are at the beginning of a pattern.

Comment (hdgarrood) [reply System/FilePath/Glob/Base.hs]: With this code, go will convert `ExtSeparator` into `Literal '.'` at the very start of a pattern, and then `checkUnmatchable` will convert that into just `[Unmatchable]`, so any pattern starting with a dot, whether explicit or not, becomes `[Unmatchable]`. I think we do need a separate function to deal with `ExtSeparator` at the very start, because the way `go` is used recursively means that we are not able to tell when we are at the beginning of a pattern.

Comment (Deewiant) [reply System/FilePath/Glob/Base.hs]: Would it work to make that a separate case at the very top level?

```haskell
optimize (Pattern pat) =
   Pattern . fin $
      case pat of
         ExtSeparator : ts = checkUnmatchable (Literal '.' :) (go ts)
         _ = checkUnmatchable id (go pat)

 where checkUnmatchable f ts = if isUnmatchable ts then [Unmatchable] else f ts
 -- isUnmatchable akin to the checkUnmatchable from my earlier comment but returning Bool
```

Sorry for continuing to shoot from the hip here but `convertExtSeparators` still smells overly complex to me.

Comment (hdgarrood) [reply System/FilePath/Glob/Base.hs]: No, I totally understand; this is just the least complex way I've been able to come up with so far, but I wouldn't go so far as to claim that it's the least complex way full stop. Unfortunately I'm quite busy at the moment so I might not be able to come back to this for a while.

Comment (Deewiant) [reply System/FilePath/Glob/Base.hs]: That's alright. In that case I'll merge this now, since after all it does fulfill its purpose, and try to simplify it myself when I get a spare moment.

I'll also do a new release next week, whether or not I have time to do something codewise before then. It's already been too long.

**Comment by Deewiant** [comment]
Thanks for all the effort you've put in!

**Comment by hdgarrood** [comment]
Sure thing - thanks for all your reviews! Incidentally I didn’t yet get a chance to try your most recent suggestion for simplifying that function so that would probably be a good place to start if you were going to revisit this.

**Comment by Deewiant** [comment]
Yep, that was my plan. :-)
